README.md

```diff
@@ -1,10 +1,260 @@
+by the `.send_command()` call arguments. Please see the function signature in
+the file `test-harness/modules/mocker/common.py` for details.
+
+#### Endpoints
+Endpoints are threading HTTP servers based on http.server python module. Because
+of the fact that there are many different kinds of endpoints that need to be
+employed by mocker, and that there is a lot of shared behaviour between them,
+they follow inheritance tree as below:
+
+<img src="endpoint_class_hierarchy.png" alt="" width="100%" align="middle">
+
+* `ReflectingUnixSocketEndpoint`, `ReflectingTCPIPEndpoint`: both of them send back
+  the request data in the response body to the client for inspection. The only
+  difference between them is that the former is listening on Unix Socket
+  and the latter is listening on TCP/IP socket. They are useful for very simple
+  tests that only check if the request is hitting the right upstream and the
+  headers are correct.
+* `BouncerEndpoint`, `MarathonEndpoint`, `MesosEndpoint`: specialized endpoints that
+  are mimicking Bouncer, Marathon and Mesos respectively. Apart from the basic
+  functionality (reset, bork, etc...), they are also capable of recording requests
+  sent to them, and then returning them through mockers `send_command` back to the
+  tests code.
+* all remaining endpoints depicted in the hierarchy are not directly usable but
+  can be inherited from and extended if necessary in order to meet the
+  requirements of testing code.
+
+Each endpoint has an id, that it basically it's http address (i.e.
+`http://127.0.0.1:8080` or `http:///run/dcos/dcos-metrics-agent.sock`). It's
+available via the `.id()` method. They are started and stopped via `.start()`
+and `.stop()` methods during the start and stop of mocker instance
+respectively. Each endpoint can be set to respond to each and every request
+with an error (`500 Internal server error`).
+
+#### Subprocess management
+Pytest fixtures start a couple of subprocesses:
+* two dnsmasq instances
+* the Adminrouter itself
+
+These do not always log to stderr/stdout, so a very simple syslog mock is also
+provided. All the stdouts and stderrs are piped into the central log
+processing class LogCatcher. 
+
+##### LogCatcher
+As mentioned in previous paragraph, all the stdout/stder file descriptors and
+syslog messages are piped into this class. It uses `poll()` call to monitor all
+the sources for new information and push it into:
+* standard python logging module
+* logs located in `test-harness/logs/`
+* internal buffer which can be used by tests for logging-based testing
+
+The internal buffer is available through
+`stdout_line_buffer`/`stderr_line_buffer` methods of AR object, Syslog
+object(available through a fixture), and dnsmasq processes (also through
+fixture). The buffer itself is implemented as a plain python list where each
+log line represents a single entry. This list is shared across all the objects
```

Comment:
A bit simplified answer, but still :)

> A general note: since we're using stdout/stderr for reading logs, the log's native data "type" format is a byte stream. This then seems to be get split into chunks, making it a sequence of byte sequences (each representing a "line").

Jup, readline does all the magic for PIPE objects: https://github.com/vespian/adminrouter/blob/prozlach/DCOS-11918_add_test_harness/test-harness/modules/runner/common.py#L145

> In the Python code that exposes the logs for analysis, I expect the lines to be exposed as byte sequences or as text resulting from a decoding step which is tolerant for unexpected bytes.

We do both :)

Tthe fixture and tests need to process data *as text* so `.decode` + `backslash replace`: https://github.com/vespian/adminrouter/blob/prozlach/DCOS-11918_add_test_harness/test-harness/modules/runner/common.py#L207-L209 We are guessing that the output from nginx & friends is utf8 though.

Log files are written without any conversion (pass through - binary data read from pipe is written to FD open in 'wb' mode: https://github.com/vespian/adminrouter/blob/prozlach/DCOS-11918_add_test_harness/test-harness/modules/runner/common.py#L214-L216

> I assume it's pretty much possible to make nginx emit log byte sequences that do not make sense in any codec, so let's never blindly decode those byte sequences via bytesquence.decode('utf-8'), as it can blow up at any time. Here, we need a relaxed error handler.

See above